### PR TITLE
Input system and mutability.

### PR DESCRIPTION
**Pull request by Britefury**
Separated `LSpaceArea` into `LSpaceWidget` and `LSpaceArea` in order to separate responsibilities.
Text elements are now mutable, in that their text can be changed and the relevant updates will be scheduled.

### CODE REVIEW

File: src/lib/lspace/elements/border_element.rs [hunk 1]
```diff
@@ -123,16 +124,15 @@ impl TElement for BorderElement {
...
 
 impl TContainerElement for BorderElement {
     fn children(&self) -> Ref<[ElementRef]> {
-        let m = self.m.borrow();
         Ref::map(self.m.borrow(), |m| m.bin.children())
     }
 
     fn compute_x_req(&self) -> LReq {
-        let mut mm = self.m.borrow_mut();
+        let
```
Comment (ltratt) [inline src/lib/lspace/elements/border_element.rs]: Join the lines?

File: src/lib/lspace/elements/element_ctx.rs [hunk 2]
```diff
@@ -40,3 +36,45 @@ impl ElementContext {
         };
     }
 }
+
+
```
Comment (ltratt) [inline src/lib/lspace/elements/element_ctx.rs:40]: Double blank line?

File: src/lib/lspace/elements/element_layout.rs [hunk 3]
```diff
@@ -137,4 +139,14 @@ impl ElementAlloc {
             p_ref = p.get_parent();
         }
     }
+
+    pub fn local_bbox(&self) -> BBox2 {
+        BBox2::from_lower_size(Point2::origin(),
+                               Vector2::new(self.x_alloc.actual_size(), self.y_alloc.actual_size()))
+
```
Comment (ltratt) [inline src/lib/lspace/elements/element_layout.rs]: Blank line.

File: src/lib/lspace/elements/root_element.rs [hunk 4]
```diff
@@ -114,16 +121,15 @@ impl TElement for RootElement {
     }
 }
 
-const NO_CHILDREN: [ElementRef; 0] = [];
+
```
Comment (ltratt) [inline src/lib/lspace/elements/root_element.rs]: 3 blank lines?

File: src/lib/lspace/elements/root_element.rs [hunk 5]
```diff
@@ -114,16 +121,15 @@ impl TElement for RootElement {
...
 
 impl TContainerElement for RootElement {
     fn children(&self) -> Ref<[ElementRef]> {
-        let m = self.m.borrow();
         return Ref::map(self.m.borrow(), |m| m.bin.children());
```
Comment (ltratt) [inline src/lib/lspace/elements/root_element.rs]: Remove `return`?

File: src/lib/lspace/elements/text_element.rs [hunk 6]
```diff
@@ -117,21 +124,30 @@ pub struct TextElement {
...
+    pub fn new_in_area(text: String, style: Rc<TextStyleParams>,
+                       area: &LSpaceArea) -> TextElement {
+        TextElement::new(text, style, &*area.element_context())
+    }
 }
 
 impl TElement for TextElement {
     /// Interface acquisition
+    fn as_text_element(&self) -> Option<&TTextElement> {
+        return Some(self);
```
Comment (ltratt) [inline src/lib/lspace/elements/text_element.rs]: Remove `return`?

File: src/lib/lspace/input/inputmodifier.rs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    pub fn from_values(value: u16) -> InputModifierState {
+        return InputModifierState{value: value};
+    }
+
+    pub fn from_buttons_and_keys(buttons: u8, keys: u8) -> InputModifierState {
+        let value = (buttons as u16) | ((keys as u16) << 8);
+        return InputModifierState{value: value};
+    }
+
+
```
Comment (ltratt) [inline src/lib/lspace/input/inputmodifier.rs]: Double blank line.

File: src/lib/lspace/input/inputmodifier.rs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    pub fn buttons(&self) -> u8 {
+        return (self.value & BUTTONS_MASK) as u8;
+    }
+
+    pub fn keys(&self) -> u8 {
+        return ((self.value & KEYS_MASK) >> 8) as u8;
+    }
+
+
```
Comment (ltratt) [inline src/lib/lspace/input/inputmodifier.rs]: Ditto.

File: src/lib/lspace/input/inputmodifier.rs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    pub fn test_buttons(&self, mask: u8) -> bool {
+        return (self.buttons() & mask) != 0;
+    }
+
+    pub fn test_keys(&self, mask: u8) -> bool {
+        return (self.keys() & mask) != 0;
+    }
+
```
Comment (ltratt) [inline src/lib/lspace/input/inputmodifier.rs:53]: Ditto.

File: src/lib/lspace/input/keyboard.rs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        KeyEvent{mod_state: mod_state, key_val: key_val, key_string: key_string,
+            event_type: KeyEventType::Press}
+    }
+
+    pub fn new_release(mod_state: InputModifierState, key_val: u32, key_string: String) -> KeyEvent {
+        KeyEvent{mod_state: mod_state, key_val: key_val, key_string: key_string,
+            event_type: KeyEventType::Release}
+    }
+
+
```
Comment (ltratt) [inline src/lib/lspace/input/keyboard.rs]: Double blank lines.

File: src/lib/lspace/input/keyboard.rs [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    m: RefCell<KeyboardMut>
+}
+
+impl Keyboard {
+    pub fn new() -> Keyboard {
+        return Keyboard{
+            m: RefCell::new(KeyboardMut::new())
+        };
+    }
+
```
Comment (ltratt) [inline src/lib/lspace/input/keyboard.rs:114]: Double blank line.

File: src/lib/lspace/input/keyboard.rs [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    pub fn on_key_press(&self, mod_state: InputModifierState, key_val: u32, key_string: String) {
+        self.m.borrow().on_key_press(mod_state, key_val, key_string);
+    }
+
+    pub fn on_key_release(&self, mod_state: InputModifierState, key_val: u32, key_string: String) {
+        self.m.borrow().on_key_release(mod_state, key_val, key_string);
+    }
+
+
```
Comment (ltratt) [inline src/lib/lspace/input/keyboard.rs]: Ditto.

File: src/lib/lspace/input/pointer.rs [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+pub struct Pointer {
+    position: Cell<PointerPosition>,
+}
+
+impl Pointer {
+    pub fn new() -> Pointer {
+        return Pointer{position: Cell::new(PointerPosition::out_of_bounds())};
+    }
+
```
Comment (ltratt) [inline src/lib/lspace/input/pointer.rs:60]: Double blank.

File: src/lib/lspace/input/pointer.rs [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+    pub fn set_position(&self, pos: PointerPosition) {
+        self.position.set(pos);
+    }
+}
+
+impl <'a> TPointer<'a> for Pointer {
+    fn position(&self) -> PointerPosition {
+        return self.position.get();
```
Comment (ltratt) [inline src/lib/lspace/input/pointer.rs]: Remove the `return`s from these methods?

File: src/lib/lspace/lspace_area.rs [hunk 15]
```diff
@@ -3,274 +3,232 @@
...
+        if self.layout_required || true {
+            let e = self.root_element.as_root_element().unwrap();
+            let t1 = time::precise_time_ns();
+            let layout_ctx = ElementLayoutContext::new(&self.elem_ctx, cairo_ctx);
+            let rx = e.root_requisition_x(&layout_ctx);
+            e.root_allocate_x(self.width as f64);
+            let ry = e.root_requisition_y();
+            e.root_allocate_y(ry);
+            let t2 = time::precise_time_ns();
+            println!("Layout time: {}", (t2-t1) as f64 * 1.0e-9);
```
Comment (ltratt) [inline src/lib/lspace/lspace_area.rs]: Isn't this a debugging aid that should be removed?

File: src/lib/lspace/lspace_area.rs [hunk 16]
```diff
@@ -3,274 +3,232 @@
...
-                println!("Render time: {}", (t2-t1) as f64 * 1.0e-9);
-            },
-            &None => {}
-        }
+        let e = self.root_element.as_root_element().unwrap();
+        let t1 = time::precise_time_ns();
+        e.draw(cairo_ctx, &BBox2::from_lower_size(Point2::origin(),
+                Vector2::new(self.width as f64, self.height as f64)));
+        let t2 = time::precise_time_ns();
+        println!("Render time: {}", (t2-t1) as f64 * 1.0e-9);
```
Comment (ltratt) [inline src/lib/lspace/lspace_area.rs]: Another debugging aid?

File: src/lib/lspace/lspace_widget.rs [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+use input::inputmodifier::{self, InputModifierState};
+
+use geom::point2::Point2;
+use geom::bbox2::BBox2;
+use pres::pres::Pres;
+use lspace_area::{LSpaceArea, TLSpaceListener};
+
+
+fn gdk_modifier_to_input_mod_state(gdk_state: gdk_ffi::GdkModifierType) -> InputModifierState {
+    let mut
```
Comment (ltratt) [inline src/lib/lspace/lspace_widget.rs]: Join line?

File: src/lib/lspace/lspace_widget.rs [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+            });
+        }
+
+        {
+            let state_clone = wrapped_area.clone();
+            drawing_area.connect_button_press_event(move |widget, event_button| {
+                let mod_state = gdk_modifier_to_input_mod_state(event_button.state);
+                let pos = Point2::new(event_button.x, event_button.y);
+                state_clone.on_button_press(mod_state, pos, event_button.button);
+                return Inhibit(true);
```
Comment (ltratt) [inline src/lib/lspace/lspace_widget.rs]: Remove the `return` from this and the subsequent closures?

File: src/lib/lspace/pres/pres.rs [hunk 19]
```diff
@@ -1,24 +1,16 @@
...
-    pub elem_ctx: &'a RefCell<ElementContext>,
-    pub cairo_ctx: &'a Context,
+    pub elem_ctx: &'a ElementContext,
 }
 
 impl <'a> PresBuildCtx<'a> {
-    pub fn new(elem_ctx: &'a RefCell<ElementContext>, cairo_ctx: &'a Context) -> PresBuildCtx<'a> {
-        return PresBuildCtx{elem_ctx: elem_ctx, cairo_ctx: cairo_ctx};
+    pub fn new(elem_ctx: &'a ElementContext) -> PresBuildCtx<'a> {
+        return PresBuildCtx{elem_ctx: elem_ctx};
```
Comment (ltratt) [inline src/lib/lspace/pres/pres.rs]: Remove `return`.

File: src/lib/lspace/elements/element_ctx.rs (hunk 2 above)
Comment (Britefury) [inline src/lib/lspace/elements/element_ctx.rs:40]: Its separating a couple of types.

Comment (ltratt) [inline src/lib/lspace/elements/element_ctx.rs:40]: OK.

File: src/lib/lspace/input/keyboard.rs (hunk 11 above)
Comment (Britefury) [inline src/lib/lspace/input/keyboard.rs:114]: Separating two types.

File: src/lib/lspace/lspace_area.rs (hunk 15 above)
Comment (Britefury) [inline src/lib/lspace/lspace_area.rs]: They are quite useful as they give you an idea as to current performance. Suggest we leave in while in early development.

File: src/lib/lspace/lspace_area.rs (hunk 16 above)
Comment (Britefury) [inline src/lib/lspace/lspace_area.rs]: Ditto; see above.

File: src/lib/lspace/lspace_area.rs (hunk 15 above)
Comment (ltratt) [inline src/lib/lspace/lspace_area.rs]: The problem is that we'll forget about them, which is always the fate of debugging code, which is why so many GUI programs spew vast quantities of no-longer-useful stuff to stdout. Can we enable these prints only in debug mode (with a macro or ...?)?

File: src/lib/lspace/input/keyboard.rs (hunk 12 above)
Comment (Britefury) [inline src/lib/lspace/input/keyboard.rs]: Should be fixed.

**Comment by Britefury** [comment]
Ready for merge.

**Comment by ltratt** [comment]
@Britefury Jasper hasn't finished his pass yet!

**Comment by ltratt** [comment]
@Britefury We also need to do something about the `print`ing of timings. Assuming this is necessary -- which I'm not convinced by myself, but it's something over which reasonable people can differ -- this is a debugging feature. Let's _only_ enable it in debug mode please.

**Comment by jaheba** [comment]
Have looked through it. Way to much code for a single PR.

Is there some lib in rust which implements an observer or ist it too simply to just implement so we do it ourselves?

**Comment by Britefury** [comment]
@jaheba Could you please edit your comments to indicate line numbers please?

**Comment by ltratt** [comment]
@Britefury click on the heading e.g. `jaheba commented on src/lib/lspace/lspace_widget.rs in 2a58bbd 12 minutes ago` notice that the filename is clickable: it takes you to the right point.

**Comment by jaheba** [comment]
@Britefury Shouldn't you see the lines here: https://github.com/softdevteam/lspace/commit/2a58bbd2f48015d5da0bd12ace867243dfcc4810 ?

**Comment by ltratt** [comment]
[This is a really poor bit of UI on GitHub's part IMHO, but oh well...]

**Comment by Britefury** [comment]
Gtk GDK enum usage is fixed.

**Comment by ltratt** [comment]
So I guess it's squash-then-merge time?

**Comment by Britefury** [comment]
Squashed.

**Pull request merged**